File: qt/gen_qanystringview.go
package qt

import (
	"unsafe"
)

type QAnyStringView struct {
	h          uintptr
	isSubclass bool
}

// NewQAnyStringView constructs a new QAnyStringView object.
func NewQAnyStringView() *QAnyStringView {

	ret := newQAnyStringView(QAnyStringView_new())
	ret.isSubclass = true
	return ret
}

// NewQAnyStringView2 constructs a new QAnyStringView object.
func NewQAnyStringView2(str []byte) *QAnyStringView {
	str_alias := struct_miqt_string{}
	str_alias.data = (char)(unsafe.Pointer(&str[0]))
	str_alias.len = size_t(len(str))

	ret := newQAnyStringView(QAnyStringView_new2(str_alias))
	ret.isSubclass = true
	return ret
}

// NewQAnyStringView3 constructs a new QAnyStringView object.
func NewQAnyStringView3(str string) *QAnyStringView {
	str_ms := struct_miqt_string{}
	str_ms.data = CString(str)
	str_ms.len = size_t(len(str))
	defer free(unsafe.Pointer(str_ms.data))

	ret := newQAnyStringView(QAnyStringView_new3(str_ms))
	ret.isSubclass = true
	return ret
}

// NewQAnyStringView4 constructs a new QAnyStringView object.
func NewQAnyStringView4(param1 *QAnyStringView) *QAnyStringView {

	ret := newQAnyStringView(QAnyStringView_new4(param1.cPointer()))
	ret.isSubclass = true
	return ret
}

func (this *QAnyStringView) Mid(pos int64) *QAnyStringView {
	_goptr := newQAnyStringView(QAnyStringView_Mid(this.h, (ptrdiff_t)(pos)))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QAnyStringView) Left(n int64) *QAnyStringView {
	_goptr := newQAnyStringView(QAnyStringView_Left(this.h, (ptrdiff_t)(n)))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QAnyStringView) Right(n int64) *QAnyStringView {
	_goptr := newQAnyStringView(QAnyStringView_Right(this.h, (ptrdiff_t)(n)))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QAnyStringView) Sliced(pos int64) *QAnyStringView {
	_goptr := newQAnyStringView(QAnyStringView_Sliced(this.h, (ptrdiff_t)(pos)))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QAnyStringView) Sliced2(pos int64, n int64) *QAnyStringView {
	_goptr := newQAnyStringView(QAnyStringView_Sliced2(this.h, (ptrdiff_t)(pos), (ptrdiff_t)(n)))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QAnyStringView) First(n int64) *QAnyStringView {
	_goptr := newQAnyStringView(QAnyStringView_First(this.h, (ptrdiff_t)(n)))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QAnyStringView) Last(n int64) *QAnyStringView {
	_goptr := newQAnyStringView(QAnyStringView_Last(this.h, (ptrdiff_t)(n)))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QAnyStringView) Chopped(n int64) *QAnyStringView {
	_goptr := newQAnyStringView(QAnyStringView_Chopped(this.h, (ptrdiff_t)(n)))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QAnyStringView) Slice(pos int64) *QAnyStringView {
	return newQAnyStringView(QAnyStringView_Slice(this.h, (ptrdiff_t)(pos)))
}

func (this *QAnyStringView) Slice2(pos int64, n int64) *QAnyStringView {
	return newQAnyStringView(QAnyStringView_Slice2(this.h, (ptrdiff_t)(pos), (ptrdiff_t)(n)))
}

func (this *QAnyStringView) Truncate(n int64) {
	QAnyStringView_Truncate(this.h, (ptrdiff_t)(n))
}

func (this *QAnyStringView) Chop(n int64) {
	QAnyStringView_Chop(this.h, (ptrdiff_t)(n))
}

func (this *QAnyStringView) ToString() string {
	var _ms struct_miqt_string = QAnyStringView_ToString(this.h)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QAnyStringView) Size() int64 {
	return (int64)(QAnyStringView_Size(this.h))
}

func (this *QAnyStringView) Data() unsafe.Pointer {
	return (unsafe.Pointer)(QAnyStringView_Data(this.h))
}

func QAnyStringView_Compare(lhs QAnyStringView, rhs QAnyStringView) int {
	return (int)(QAnyStringView_Compare(lhs.cPointer(), rhs.cPointer()))
}

func QAnyStringView_Equal(lhs QAnyStringView, rhs QAnyStringView) bool {
	return (bool)(QAnyStringView_Equal(lhs.cPointer(), rhs.cPointer()))
}

func (this *QAnyStringView) Front() *QChar {
	_goptr := newQChar(QAnyStringView_Front(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QAnyStringView) Back() *QChar {
	_goptr := newQChar(QAnyStringView_Back(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QAnyStringView) Empty() bool {
	return (bool)(QAnyStringView_Empty(this.h))
}

func (this *QAnyStringView) SizeBytes() int64 {
	return (int64)(QAnyStringView_SizeBytes(this.h))
}

func (this *QAnyStringView) MaxSize() int64 {
	return (int64)(QAnyStringView_MaxSize(this.h))
}

func (this *QAnyStringView) IsNull() bool {
	return (bool)(QAnyStringView_IsNull(this.h))
}

func (this *QAnyStringView) IsEmpty() bool {
	return (bool)(QAnyStringView_IsEmpty(this.h))
}

func (this *QAnyStringView) Length() int64 {
	return (int64)(QAnyStringView_Length(this.h))
}

func (this *QAnyStringView) Mid2(pos int64, n int64) *QAnyStringView {
	_goptr := newQAnyStringView(QAnyStringView_Mid2(this.h, (ptrdiff_t)(pos), (ptrdiff_t)(n)))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func QAnyStringView_Compare3(lhs QAnyStringView, rhs QAnyStringView, cs CaseSensitivity) int {
	return (int)(QAnyStringView_Compare3(lhs.cPointer(), rhs.cPointer(), (int)(cs)))
}

File: qt/gen_qanystringview.cpp
// +build ignore

#include <QAnyStringView>
#include <QByteArray>
#include <QChar>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qanystringview.h>
#include "gen_qanystringview.h"

QAnyStringView* QAnyStringView_new() {
	return new QAnyStringView();
}

QAnyStringView* QAnyStringView_new2(struct miqt_string str) {
	QByteArray str_QByteArray(str.data, str.len);
	return new QAnyStringView(str_QByteArray);
}

QAnyStringView* QAnyStringView_new3(struct miqt_string str) {
	QString str_QString = QString::fromUtf8(str.data, str.len);
	return new QAnyStringView(str_QString);
}

QAnyStringView* QAnyStringView_new4(QAnyStringView* param1) {
	return new QAnyStringView(*param1);
}

QAnyStringView* QAnyStringView_Mid(const QAnyStringView* self, ptrdiff_t pos) {
	return new QAnyStringView(self->mid((qsizetype)(pos)));
}

QAnyStringView* QAnyStringView_Left(const QAnyStringView* self, ptrdiff_t n) {
	return new QAnyStringView(self->left((qsizetype)(n)));
}

QAnyStringView* QAnyStringView_Right(const QAnyStringView* self, ptrdiff_t n) {
	return new QAnyStringView(self->right((qsizetype)(n)));
}

QAnyStringView* QAnyStringView_Sliced(const QAnyStringView* self, ptrdiff_t pos) {
	return new QAnyStringView(self->sliced((qsizetype)(pos)));
}

QAnyStringView* QAnyStringView_Sliced2(const QAnyStringView* self, ptrdiff_t pos, ptrdiff_t n) {
	return new QAnyStringView(self->sliced((qsizetype)(pos), (qsizetype)(n)));
}

QAnyStringView* QAnyStringView_First(const QAnyStringView* self, ptrdiff_t n) {
	return new QAnyStringView(self->first((qsizetype)(n)));
}

QAnyStringView* QAnyStringView_Last(const QAnyStringView* self, ptrdiff_t n) {
	return new QAnyStringView(self->last((qsizetype)(n)));
}

QAnyStringView* QAnyStringView_Chopped(const QAnyStringView* self, ptrdiff_t n) {
	return new QAnyStringView(self->chopped((qsizetype)(n)));
}

QAnyStringView* QAnyStringView_Slice(QAnyStringView* self, ptrdiff_t pos) {
	QAnyStringView& _ret = self->slice((qsizetype)(pos));
	// Cast returned reference into pointer
	return &_ret;
}

QAnyStringView* QAnyStringView_Slice2(QAnyStringView* self, ptrdiff_t pos, ptrdiff_t n) {
	QAnyStringView& _ret = self->slice((qsizetype)(pos), (qsizetype)(n));
	// Cast returned reference into pointer
	return &_ret;
}

void QAnyStringView_Truncate(QAnyStringView* self, ptrdiff_t n) {
	self->truncate((qsizetype)(n));
}

void QAnyStringView_Chop(QAnyStringView* self, ptrdiff_t n) {
	self->chop((qsizetype)(n));
}

struct miqt_string QAnyStringView_ToString(const QAnyStringView* self) {
	QString _ret = self->toString();
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

ptrdiff_t QAnyStringView_Size(const QAnyStringView* self) {
	qsizetype _ret = self->size();
	return static_cast<ptrdiff_t>(_ret);
}

const void* QAnyStringView_Data(const QAnyStringView* self) {
	return (const void*) self->data();
}

int QAnyStringView_Compare(QAnyStringView* lhs, QAnyStringView* rhs) {
	return QAnyStringView::compare(*lhs, *rhs);
}

bool QAnyStringView_Equal(QAnyStringView* lhs, QAnyStringView* rhs) {
	return QAnyStringView::equal(*lhs, *rhs);
}

QChar* QAnyStringView_Front(const QAnyStringView* self) {
	return new QChar(self->front());
}

QChar* QAnyStringView_Back(const QAnyStringView* self) {
	return new QChar(self->back());
}

bool QAnyStringView_Empty(const QAnyStringView* self) {
	return self->empty();
}

ptrdiff_t QAnyStringView_SizeBytes(const QAnyStringView* self) {
	qsizetype _ret = self->size_bytes();
	return static_cast<ptrdiff_t>(_ret);
}

ptrdiff_t QAnyStringView_MaxSize(const QAnyStringView* self) {
	qsizetype _ret = self->max_size();
	return static_cast<ptrdiff_t>(_ret);
}

bool QAnyStringView_IsNull(const QAnyStringView* self) {
	return self->isNull();
}

bool QAnyStringView_IsEmpty(const QAnyStringView* self) {
	return self->isEmpty();
}

ptrdiff_t QAnyStringView_Length(const QAnyStringView* self) {
	qsizetype _ret = self->length();
	return static_cast<ptrdiff_t>(_ret);
}

QAnyStringView* QAnyStringView_Mid2(const QAnyStringView* self, ptrdiff_t pos, ptrdiff_t n) {
	return new QAnyStringView(self->mid((qsizetype)(pos), (qsizetype)(n)));
}

int QAnyStringView_Compare3(QAnyStringView* lhs, QAnyStringView* rhs, int cs) {
	return QAnyStringView::compare(*lhs, *rhs, static_cast<Qt::CaseSensitivity>(cs));
}

void QAnyStringView_Delete(QAnyStringView* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<QAnyStringView*>( self );
	} else {
		delete self;
	}
}

